### PR TITLE
Initial implementation of Pagerduty support

### PR DESCRIPTION
**Pull request by cewood**
Still a bit rough around the edges, and need to rework the Twillio
integration to the new way of things.

### CODE REVIEW

**Comment by cewood** [comment]
So this expects a new format for the config.json file, which should look like so:

```
{
    "handler": {
        "pagerduty": {
            "some-config": {
                "integration-key": "some secret here"
            }
        },
        "twilio": {
            "another-config": {
                "reply_number": "+49 123 4567 8902",
                "account_sid": "some secret here",
                "auth_token": "some secret here",
                "url_str": "https://api.twilio.com/2010-04-01/Accounts/ladida/Messages.json",
                "reply_handler_port": 8080,
                "reply_handler_callback": "/handler/replies",
                "recipients": {
                    "john doe": "+49 123 4567 8901"
                }
            }
        }
    },
    "test_interval": 3
}
```

And for the services.json file, it should be like so:

```
[
    {
        "name": "Fake dummy endpoint",
        "url": "https://thisendpointdoesntexist.io/please/fail/me",
        "disabled": false,
        "tests": {
            "min_response_size": 23,
            "max_response_time": 1500,
            "json": true,
            "cert": false
        },
        "handlers": [
            {
                "type": "pagerduty",
                "name": "some-config"
            },
            {
                "type": "twilio",
                "name": "another-config"
            }
        ]
    }
] 
```

File: alerting/handler.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package alerting
+
+import (
+	"github.com/datajet-io/peekaboo/services"
+)
+
+type Handler interface {
+	TriggerAlert(service services.Service, alert Alert, details map[string]interface{}) error
+	Print()
```
Comment (nchern) [inline alerting/handler.go]: what is the purpose of this function?

File: alerting/messaging.go [hunk 2]
```diff
@@ -1,153 +1,154 @@
...
-
-//RunReplyHandler starts accepting responses
-func (m *Messaging) RunReplyHandler() {
-	http.HandleFunc(m.config.ReplyHandlerCallback, m.replyHandler)
-
-	port := strconv.Itoa(m.config.ReplyHandlerPort)
-
-	http.ListenAndServe(":"+port, nil)
-}
+//
```
Comment (nchern) [inline alerting/messaging.go]: if we do not need this, lets just remove. You can do it in a separate commit, so it will be easy to revert.

File: alerting/pagerduty.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	"io/ioutil"
+	"net/http"
+
+	"github.com/datajet-io/peekaboo/services"
+)
+
+const apiEndpoint = "https://events.pagerduty.com/generic/2010-04-15/create_event.json"
+
+//Pagerduty represents the configuration of the Pageruty account used for notifications
+type Pagerduty struct {
```
Comment (nchern) [inline alerting/pagerduty.go]: it is not just a configuration, it looks more like a client, doesn't it? lets call it smth like PagerDutyClient

File: alerting/handler.go [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package alerting
+
+import (
+	"github.com/datajet-io/peekaboo/services"
+)
+
+type Handler interface {
```
Comment (nchern) [inline alerting/handler.go]: Its is smth more like Alerter/AlertTrigger etc... Handler is super generic

File: main.go [hunk 5]
```diff
@@ -52,32 +51,32 @@ func hasInternet(logger zap.Logger) bool {
...
-		if numberOfServices == 1 {
-			t = "service"
-		}
-
-		m := fmt.Sprintf("Hi %s! You will receive alerts for %d %s on Peekaboo - the messaging-based health checker. Stop alerts by texting 'stop <ID>', start alerts with 'start <ID>'", s[1], numberOfServices, t)
-
-		msgs.SendSMS(s[0], m)
-	}
-}
+// func welcomeOwners(srvs *services.Services, msgs *alerting.Messaging) {
```
Comment (nchern) [inline main.go]: same here - lets remove if we does not need.

File: services/services.go [hunk 6]
```diff
@@ -52,22 +52,22 @@ func (s *Services) Get(id string) (*Service, error) {
...
-			if owner.Cell == cell {
-				return &owner, nil
-			}
-
-		}
-	}
-
-	return nil, errors.New("No owner found with the cell number." + cell)
-}
+// // GetOwnerByCell returns the service with the given id
```
Comment (nchern) [inline services/services.go]: same here

File: services/service.go [hunk 7]
```diff
@@ -13,10 +13,10 @@ import (
 	"github.com/uber-go/zap"
 )
 
-//Owner represents the owner of a service who will be contacted in case of test failure
-type Owner struct {
+//Handler contains which notifiers to use for this service if it alerts
+type Handler struct {
```
Comment (nchern) [inline services/service.go]: so this is smth like alert handler as well?

File: alerting/handler.go [hunk 8]
```diff
@@ -0,0 +1,9 @@
+package alerting
+
+import (
+	"github.com/datajet-io/peekaboo/services"
+)
+
+type AlertHandler interface {
```
Comment (nchern) [inline alerting/handler.go]: If it triggers an alert, its not a handler its more like a trigger? (now this entity conflicts a bit with entity in service code below).

File: alerting/pagerduty.go [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+	Description  string                 `json:"description"`
+	Client       string                 `json:"client"`
+	Details      map[string]interface{} `json:"details"`
+}
+
+func NewPagerdutyClient(key string) *PagerdutyClient {
+	return &PagerdutyClient{integrationKey: key}
+}
+
+func (p PagerdutyClient) Print() {
```
Comment (nchern) [inline alerting/pagerduty.go]: if this is not used lets just delete it

File: alerting/pagerduty.go [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+			"Error marshaling 'eventData' to JSON",
+			zap.Object("eventData", eventData),
+			zap.Error(err),
+		)
+		return err
+	}
+
+	// Make request, with retrying support
+	operation := func() error {
+		req, err := http.NewRequest("POST", apiEndpoint, bytes.NewBuffer(eventDataJson))
```
Comment (nchern) [inline alerting/pagerduty.go]: in case of an error here, should we return?seems we can not proceed

File: alerting/pagerduty.go [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+		resp, err := client.Do(req)
+		if err != nil {
+			service.Logger.Warn(
+				"Encountered error calling Pagerduty API",
+				zap.Error(err),
+			)
+			return err
+		}
+		defer resp.Body.Close()
+		respBody, err := ioutil.ReadAll(resp.Body)
```
Comment (nchern) [inline alerting/pagerduty.go:79]: again in case of an error, we should return(or handle correctly)

File: alerting/pagerduty.go [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+		statusCode := resp.StatusCode
+		if statusCode == 200 {
```
Comment (nchern) [inline alerting/pagerduty.go:89]: this looks strange. 
1. err variable here is related for previous call and should be handled before(prev comment).
2. What is the app logic  here -we return an error in case of HTTP OK?

File: services/service.go (hunk 7 above)
Comment (cewood) [reply services/service.go]: DONE

File: alerting/handler.go (hunk 8 above)
Comment (cewood) [reply alerting/handler.go]: DONE

File: alerting/pagerduty.go (hunk 9 above)
Comment (cewood) [reply alerting/pagerduty.go]: DONE

File: alerting/pagerduty.go (hunk 10 above)
Comment (cewood) [reply alerting/pagerduty.go]: DONE

Couldn't return since we're in a retrying block, but I moved these setup tasks outside the retried section, so it will immediately fail in that case and skip the retries. Good catch.

File: alerting/pagerduty.go (hunk 11 above)
Comment (cewood) [reply alerting/pagerduty.go:79]: DONE

File: alerting/pagerduty.go (hunk 12 above)
Comment (cewood) [reply alerting/pagerduty.go:89]: DONE

You're right, it's not clear. Yes the function returns an error, or nil in the case of the 200 signifying success. I've replaced it correctly with an explicit `nil` value.

File: alerting/pagerduty.go [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+			return err
+		}
+
+		statusCode := resp.StatusCode
+		if statusCode == 200 {
+			return nil
+		}
+
+		switch statusCode {
+		case 400:
```
Comment (nchern) [inline alerting/pagerduty.go:94]: fyi: all these codes have constants in http package. Here we can leave this as it is for sure.

File: alerting/pagerduty.go [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+	if err != nil {
+		service.Logger.Warn(
+			"Error marshaling 'eventData' to JSON",
+			zap.Object("eventData", eventData),
+			zap.Error(err),
+		)
+		return err
+	}
+
+	req, err := http.NewRequest("POST", apiEndpoint, bytes.NewBuffer(eventDataJson))
```
Comment (nchern) [inline alerting/pagerduty.go:56]: still need a check

**Review by nchern — APPROVED** [review]
